pkg/sfu/interceptor/rtx.go

```diff
@@ -0,0 +1,173 @@
+package interceptor
+
+import (
+	"strings"
+	"sync"
+
+	"github.com/pion/interceptor"
+	"github.com/pion/sdp/v3"
+	"github.com/pion/webrtc/v3"
+
+	"github.com/livekit/livekit-server/pkg/sfu/utils"
+	"github.com/livekit/protocol/logger"
+)
+
+const (
+	SDESRepairRTPStreamIDURI = "urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id"
+
+	rtxProbeCount = 10
+)
+
+type streamInfo struct {
+	mid  string
+	rid  string
+	rsid string
+}
+
+type RTXInfoExtractorFactory struct {
+	onVideoStreamFound func(*interceptor.StreamInfo)
+	onRTXPairFound     func(repair, base uint32)
+	lock               sync.Mutex
+	streams            map[uint32]streamInfo
+	logger             logger.Logger
+}
+
+func NewRTXInfoExtractorFactory(onVideoStreamFound func(*interceptor.StreamInfo), onRTXPairFound func(repair, base uint32), logger logger.Logger) *RTXInfoExtractorFactory {
+	return &RTXInfoExtractorFactory{
+		onVideoStreamFound: onVideoStreamFound,
+		onRTXPairFound:     onRTXPairFound,
+		streams:            make(map[uint32]streamInfo),
+		logger:             logger,
+	}
+}
+
+func (f *RTXInfoExtractorFactory) NewInterceptor(id string) (interceptor.Interceptor, error) {
+	return &RTXInfoExtractor{
+		factory: f,
+		logger:  f.logger,
+	}, nil
+}
+
+func (f *RTXInfoExtractorFactory) setStreamInfo(ssrc uint32, mid, rid, rsid string) {
+	var repairSsrc, baseSsrc uint32
+	f.lock.Lock()
+
+	if rsid != "" {
+		// repair stream found, find base stream
+		for base, info := range f.streams {
+			if info.mid == mid && info.rid == rsid {
+				repairSsrc = ssrc
+				baseSsrc = base
+				delete(f.streams, base)
+				break
+			}
+		}
+	} else {
+		// base stream found, find repair stream
+		for repair, info := range f.streams {
+			if info.mid == mid && info.rsid == rid {
+				repairSsrc = repair
+				baseSsrc = ssrc
+				delete(f.streams, repair)
+				break
+			}
+		}
+	}
+
+	// no rtx pair found, save it for later
+	if repairSsrc == 0 || baseSsrc == 0 {
+		f.streams[ssrc] = streamInfo{
+			mid:  mid,
+			rid:  rid,
+			rsid: rsid,
+		}
+	}
+
+	f.lock.Unlock()
+
+	if repairSsrc != 0 && baseSsrc != 0 {
+		f.onRTXPairFound(repairSsrc, baseSsrc)
+	}
+}
+
+type RTXInfoExtractor struct {
+	interceptor.NoOp
+
+	factory *RTXInfoExtractorFactory
+	logger  logger.Logger
+}
+
+func (u *RTXInfoExtractor) BindRemoteStream(info *interceptor.StreamInfo, reader interceptor.RTPReader) interceptor.RTPReader {
+	if !strings.HasPrefix(info.MimeType, "video") {
```

Comment:
should we restrict this to `video`? Okay for now, but if possible, would be good to just keep it generic as audio NACKs are potentially use the RTX stream too.

Reply:
No need to restrict it here, move to `setTWCC` callback